streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -229,37 +230,35 @@ public Subscription subscription(Set<String> topics) {
         if (internalTopicManager != null) {
             log.debug("Starting to validate internal source topics in partition assignor.");
 
-            if (internalSourceTopicToTaskIds != null) {
-                for (Map.Entry<String, Set<TaskId>> entry : internalSourceTopicToTaskIds.entrySet()) {
-                    String topic = streamThread.jobId + "-" + entry.getKey();
+            for (Map.Entry<String, Set<TaskId>> entry : internalSourceTopicToTaskIds.entrySet()) {
+                String topic = streamThread.jobId + "-" + entry.getKey();
 
-                    // should have size 1 only
-                    int numPartitions = -1;
-                    for (TaskId task : entry.getValue()) {
-                        numPartitions = task.partition;
-                    }
+                // should have size 1 only
+                int numPartitions = -1;
+                for (TaskId task : entry.getValue()) {
+                    numPartitions = task.partition;
+                }
 
-                    internalTopicManager.makeReady(topic, numPartitions);
+                internalTopicManager.makeReady(topic, numPartitions);
 
-                    // wait until the topic metadata has been propagated to all brokers
-                    List<PartitionInfo> partitions;
-                    do {
-                        partitions = streamThread.restoreConsumer.partitionsFor(topic);
-                    } while (partitions == null || partitions.size() != numPartitions);
+                // wait until the topic metadata has been propagated to all brokers
+                List<PartitionInfo> partitions;
+                do {
+                    partitions = streamThread.restoreConsumer.partitionsFor(topic);
+                } while (partitions == null || partitions.size() != numPartitions);
 
-                    metadata.update(topic, partitions);
-                }
+                metadata.update(topic, partitions);
             }
 
             log.info("Completed validating internal source topics in partition assignor.");
         }
+        internalSourceTopicToTaskIds.clear();
```

Comment:
Yes that is intended: the original bug was that internal source topics (from repartition, for example) need to be created before we can user-code PartitionGrouper.groupPartitions(), and internalSourceTopicToTaskIds is just reused to store the num.partitions info; after that, we need to clear it so that after groupPartitions is called, it can be re-populated with user defined tasks.